coding/python_testing.rst

```diff
@@ -4,6 +4,204 @@ Python Unit Testing
+* Test classes must, ultimately, inherit from :class:`unittest.TestCase` in order to be discovered and it is :ref:`recommended <py-utils-tests>` that :lclass:`lsst.utils.tests.TestCase` be used as the base class when :lmod:`~lsst.afw` objects are involved.
+  The tests themselves must be methods of the test class with names that begin with ``test``.
+  All other methods and classes will be ignored by the test system but can be used by tests.
+* If a test method completes the test passes, if it throws an uncaught exception the test has failed.
+
+Using a more advanced test runner
+=================================
+
+All tests should be written such that they are runnable using `pytest`_, a policy adopted in :jira:`RFC-69`.
+`pytest`_ provides a much richer execution and reporting environment for tests and can be used to run multiple tests files together.
+`pytest`_ test discovery is much more flexible than that provided by :mod:`unittest` but LSST test files should not take advantage of that flexibility as it can lead to inconsistency in test reports that depend on the specific test runner.
+In particular, care must be taken not to have free functions that use a ``test`` prefix or non-\ :class:`~unittest.TestCase` test classes that are named with a ``Test`` prefix in the test files.
+
+.. note::
+   The normal way to run tests is to use :command:`scons` or :command:`scons tests`.
+   :command:`scons` will then execute each of the tests and report any failures.
+   Currently tests are executed one at a time using :command:`python` and checking the command exit status.
+   Eventually :lmod:`~lsst.sconsUtils` will be modified to call :command:`py.test` directly with all Python test files (possibly with a different order each invocation).
+   It is therefore important that during this transition period developers check that tests run correctly with `pytest`_ by explicitly running :command:`py.test` when creating or updating test files.
+
+`pytest`_ can run tests from more than one file in a single invocation and this can be used to verify that there is no state contaminating later tests.
+To run `pytest`_ use the :command:`py.test` executable:
+
+.. code-block:: shell
+
+   $ py.test tests/*.py
+
+and to ensure that the order of test execution does not matter it is useful to sometimes run the tests in reverse order:
+
+.. code-block:: shell
+
+   $ py.test `ls -r tests/*.py`
+
+When writing tests it is important that tests are skipped using the proper :mod:`unittest` :ref:`skipping framework <python:unittest-skipping>` rather than returning from the test early.
+:mod:`unittest` supports skipping of individual tests and entire classes using decorators or skip exceptions.
+LSST code sometimes raises skip exceptions in :meth:`~unittest.TestCase.setUp` or :meth:`~unittest.TestCase.setUpClass` class methods and :mod:`unittest` provides much flexibility in how skipping can be implemented.
```

Comment:
The part including and after "and unittest" seems somewhat redundant and could be dropped.